torchdata/dataloader2/reading_service.py

```diff
@@ -381,28 +447,21 @@ def _pause(self):
                 continue
             if hasattr(dp, "pause") and callable(dp.pause):
                 dp.pause()
-        if self.num_workers > 0:
-            self._worker_consumer_datapipe.request_pause()  # type: ignore[union-attr]
-        else:
-            raise RuntimeError(
-                "If you would like to use `pause` with `MultiProcessingReadingService`, "
-                "please use more than 0 worker."
-            )
+        self._worker_consumer_datapipe.request_pause()  # type: ignore[union-attr]
 
     def _resume(self):
         """
         Resumes DataPipes' activities. This is required to be called after `_pause` before
         the DataLoader can keep yielding elements.
         """
-        if self.num_workers > 0:
-            self._worker_consumer_datapipe.request_resume()  # type: ignore[union-attr]
-        else:
-            raise RuntimeError(
-                "If you would like to use `resume` with `MultiProcessingReadingService`, "
-                "please use more than 0 worker."
-            )
-        if self.main_prefetch_cnt > 0 and self.num_workers > 0:
-            self._main_prefetch_datapipe.resume()  # type: ignore[union-attr]
+        assert self._end_datapipe is not None
+
+        self._worker_consumer_datapipe.request_resume()  # type: ignore[union-attr]
+        dp_list = list_dps(traverse_dps(self._end_datapipe))
+        # Reversed order
+        for dp in dp_list[::-1]:
+            if hasattr(dp, "resume") and callable(dp.resume):
+                dp.resume()
```

Comment:
Do we need to duplicate this logic when it also exists within `iter.py` where it checks `elif isinstance(request, communication.messages.ResumeRequest):`? Similar with `pause`. 

I need more time to think but perhaps there is a way for it to only exist in one place?

Reply:
We do because the logic in `iter` is called either in worker process or dispatching process. But, this part is called in the main process for the rest of `DataPipes` remaining in the main process.

Reply:
In that case, we need something similar for `_pause` right above as well?

Reply:
I think `_pause` has already done that.